REPL crashing on standalone `_root_` reference

## Compiler version

3.3.2-RC1-bin-20230601-8814760-NIGHTLY - prints compiler stack trace, keeps REPL alive
3.3.1-RC1 - prints compiler stack trace, keeps REPL alive
3.3.0 - crashes REPL

## Minimized code

```scala
scala> _root_
```

## Output (click arrow to expand)
<details>

```scala
Warning: mocking up superclass for module class <root>
java.lang.AssertionError: assertion failed: Cannot create ClassBType from NoSymbol
	at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
	at dotty.tools.backend.jvm.BTypesFromSymbols.classBTypeFromSymbol(BTypesFromSymbols.scala:51)
	at dotty.tools.backend.jvm.BTypesFromSymbols.buildNestedInfo(BTypesFromSymbols.scala:218)
	at dotty.tools.backend.jvm.BTypesFromSymbols.setClassInfo(BTypesFromSymbols.scala:176)
	at dotty.tools.backend.jvm.BTypesFromSymbols.classBTypeFromSymbol$$anonfun$1(BTypesFromSymbols.scala:64)
	at scala.collection.mutable.HashMap.getOrElse(HashMap.scala:436)
	at dotty.tools.backend.jvm.BTypesFromSymbols.classBTypeFromSymbol(BTypesFromSymbols.scala:65)
	at dotty.tools.backend.jvm.BCodeHelpers$BCInnerClassGen.getClassBType(BCodeHelpers.scala:177)
	at dotty.tools.backend.jvm.BCodeHelpers$BCInnerClassGen.getClassBType$(BCodeHelpers.scala:129)
	at dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.getClassBType(BCodeSkelBuilder.scala:74)
	at dotty.tools.backend.jvm.BCodeHelpers.primitiveOrClassToBType$1$$anonfun$1(BCodeHelpers.scala:714)
	at scala.collection.immutable.HashMap.getOrElse(HashMap.scala:683)
	at dotty.tools.backend.jvm.BCodeHelpers.primitiveOrClassToBType$1(BCodeHelpers.scala:714)
	at dotty.tools.backend.jvm.BCodeHelpers.dotty$tools$backend$jvm$BCodeHelpers$$typeToTypeKind(BCodeHelpers.scala:733)
	at dotty.tools.backend.jvm.BCodeHelpers$BCInnerClassGen.toTypeKind(BCodeHelpers.scala:201)
	at dotty.tools.backend.jvm.BCodeHelpers$BCInnerClassGen.toTypeKind$(BCodeHelpers.scala:129)
	at dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.toTypeKind(BCodeSkelBuilder.scala:74)
	at dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.symInfoTK(BCodeSkelBuilder.scala:103)
	at dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.addClassFields$$anonfun$2(BCodeSkelBuilder.scala:352)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.addClassFields(BCodeSkelBuilder.scala:358)
	at dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.genPlainClass(BCodeSkelBuilder.scala:232)
	at dotty.tools.backend.jvm.CodeGen.genClass(CodeGen.scala:138)
	at dotty.tools.backend.jvm.CodeGen.genClassDef$1(CodeGen.scala:55)
	at dotty.tools.backend.jvm.CodeGen.genClassDefs$1(CodeGen.scala:101)
	at dotty.tools.backend.jvm.CodeGen.genClassDefs$1$$anonfun$1(CodeGen.scala:99)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.backend.jvm.CodeGen.genClassDefs$1(CodeGen.scala:99)
	at dotty.tools.backend.jvm.CodeGen.genUnit(CodeGen.scala:104)
	at dotty.tools.backend.jvm.GenBCode.run(GenBCode.scala:76)
	at dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:327)
	at scala.collection.immutable.List.map(List.scala:246)
	at dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:331)
	at dotty.tools.backend.jvm.GenBCode.runOn(GenBCode.scala:87)
	at dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:246)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1321)
	at dotty.tools.dotc.Run.runPhases$1(Run.scala:262)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:270)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$adapted$1(Run.scala:279)
	at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:67)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:279)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:200)
	at dotty.tools.repl.ReplCompiler.compile(ReplCompiler.scala:87)
	at dotty.tools.repl.ReplDriver.compile(ReplDriver.scala:307)
	at dotty.tools.repl.ReplDriver.interpret(ReplDriver.scala:269)
	at dotty.tools.repl.ReplDriver.loop$1(ReplDriver.scala:169)
	at dotty.tools.repl.ReplDriver.runUntilQuit$$anonfun$1(ReplDriver.scala:172)
	at dotty.tools.repl.ReplDriver.withRedirectedOutput(ReplDriver.scala:211)
	at dotty.tools.repl.ReplDriver.runBody$$anonfun$1(ReplDriver.scala:185)
	at dotty.tools.runner.ScalaClassLoader$.asContext(ScalaClassLoader.scala:80)
	at dotty.tools.repl.ReplDriver.runBody(ReplDriver.scala:185)
	at dotty.tools.repl.ReplDriver.runUntilQuit(ReplDriver.scala:172)
	at dotty.tools.repl.ReplDriver.tryRunning(ReplDriver.scala:134)
	at dotty.tools.repl.Main$.main(Main.scala:7)
	at dotty.tools.repl.Main.main(Main.scala)
Error while emitting rs$line$1
assertion failed: Cannot create ClassBType from NoSymbol
1 error foun
```
</details>

Also code completions on `_root_.<TAB>` in REPL don't work

--- a/compiler/src/dotty/tools/dotc/parsing/Parsers.scala
+++ b/compiler/src/dotty/tools/dotc/parsing/Parsers.scala
@@ -1072,8 +1072,15 @@ object Parsers {
       }
 
     /** Accept identifier and return Ident with its name as a term name. */
-    def termIdent(): Ident =
+    def rawTermIdent(): Ident =
       makeIdent(in.token, in.offset, ident())
+
+    /** Call `rawTermIdent`, and check it isn't a root package name. */
+    def termIdent(): Ident =
+      val ident = rawTermIdent()
+      if ident.name == nme.ROOTPKG then
+        syntaxError(em"Illegal use of root package name.")
+      ident
 
     /** Accept identifier and return Ident with its name as a type name. */
     def typeIdent(): Ident =
@@ -1112,7 +1119,7 @@ object Parsers {
      *              |  [id ‘.’] ‘this’
      *              |  [id ‘.’] ‘super’ [ClassQualifier] ‘.’ id
      */
-    def simpleRef(): Tree =
+    def simpleRef(allowRoot: Boolean = true): Tree =
       val start = in.offset
 
       def handleThis(qual: Ident) =
@@ -1129,7 +1136,7 @@ object Parsers {
       if in.token == THIS then handleThis(EmptyTypeIdent)
       else if in.token == SUPER then handleSuper(EmptyTypeIdent)
       else
-        val t = termIdent()
+        val t = if allowRoot then rawTermIdent() else termIdent()
         if in.token == DOT then
           def qual = cpy.Ident(t)(t.name.toTypeName)
           in.lookahead.token match
@@ -2965,7 +2972,7 @@ object Parsers {
      */
     def simplePattern(): Tree = in.token match {
       case IDENTIFIER | BACKQUOTED_IDENT | THIS | SUPER =>
-        simpleRef() match
+        simpleRef(allowRoot = false) match
           case id @ Ident(nme.raw.MINUS) if isNumericLit => literal(startOffset(id))
           case t => simplePatternRest(t)
       case USCORE =>

--- a/compiler/src/dotty/tools/dotc/parsing/Parsers.scala
+++ b/compiler/src/dotty/tools/dotc/parsing/Parsers.scala
@@ -1072,15 +1072,11 @@ object Parsers {
       }
 
     /** Accept identifier and return Ident with its name as a term name. */
-    def rawTermIdent(): Ident =
-      makeIdent(in.token, in.offset, ident())
-
-    /** Call `rawTermIdent`, and check it isn't a root package name. */
     def termIdent(): Ident =
-      val ident = rawTermIdent()
-      if ident.name == nme.ROOTPKG then
+      val t = makeIdent(in.token, in.offset, ident())
+      if t.name == nme.ROOTPKG then
         syntaxError(em"Illegal use of root package name.")
-      ident
+      t
 
     /** Accept identifier and return Ident with its name as a type name. */
     def typeIdent(): Ident =
@@ -1119,7 +1115,7 @@ object Parsers {
      *              |  [id ‘.’] ‘this’
      *              |  [id ‘.’] ‘super’ [ClassQualifier] ‘.’ id
      */
-    def simpleRef(allowRoot: Boolean = true): Tree =
+    def simpleRef(): Tree =
       val start = in.offset
 
       def handleThis(qual: Ident) =
@@ -1135,19 +1131,21 @@ object Parsers {
 
       if in.token == THIS then handleThis(EmptyTypeIdent)
       else if in.token == SUPER then handleSuper(EmptyTypeIdent)
-      else
-        val t = if allowRoot then rawTermIdent() else termIdent()
-        if in.token == DOT then
-          def qual = cpy.Ident(t)(t.name.toTypeName)
-          in.lookahead.token match
-            case THIS =>
-              in.nextToken()
-              handleThis(qual)
-            case SUPER =>
-              in.nextToken()
-              handleSuper(qual)
-            case _ => t
-        else t
+      else if in.lookahead.token == DOT then
+        val tok    = in.token
+        val offset = in.offset
+        val name   = ident()
+        def qual   = makeIdent(tok, offset, name.toTypeName)
+        in.lookahead.token match
+          case THIS =>
+            in.nextToken()
+            handleThis(qual)
+          case SUPER =>
+            in.nextToken()
+            handleSuper(qual)
+          case _ =>
+            makeIdent(tok, offset, name)
+      else termIdent()
     end simpleRef
 
     /** MixinQualifier ::= `[' id `]'
@@ -2972,7 +2970,7 @@ object Parsers {
      */
     def simplePattern(): Tree = in.token match {
       case IDENTIFIER | BACKQUOTED_IDENT | THIS | SUPER =>
-        simpleRef(allowRoot = false) match
+        simpleRef() match
           case id @ Ident(nme.raw.MINUS) if isNumericLit => literal(startOffset(id))
           case t => simplePatternRest(t)
       case USCORE =>

--- a/compiler/src/dotty/tools/dotc/parsing/Parsers.scala
+++ b/compiler/src/dotty/tools/dotc/parsing/Parsers.scala
@@ -1131,7 +1131,7 @@ object Parsers {
 
       if in.token == THIS then handleThis(EmptyTypeIdent)
       else if in.token == SUPER then handleSuper(EmptyTypeIdent)
-      else if in.lookahead.token == DOT then
+      else if in.token != INTERPOLATIONID && in.lookahead.token == DOT then
         val tok    = in.token
         val offset = in.offset
         val name   = ident()

--- a/compiler/src/dotty/tools/dotc/typer/Typer.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Typer.scala
@@ -583,8 +583,8 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
         if foundUnderScala2.exists && !(foundUnderScala2 =:= found) then
           report.migrationWarning(
             em"""Name resolution will change.
-                | currently selected                          : $foundUnderScala2
-                | in the future, without -source 3.0-migration: $found""", tree.srcPos)
+              | currently selected                          : $foundUnderScala2
+              | in the future, without -source 3.0-migration: $found""", tree.srcPos)
           foundUnderScala2
         else found
       finally

--- a/compiler/src/dotty/tools/dotc/typer/Typer.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Typer.scala
@@ -569,7 +569,9 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
     // optimization, it also avoids forcing imports thus potentially avoiding
     // cyclic references.
     if (name == nme.ROOTPKG)
-      return tree.withType(defn.RootPackage.termRef)
+      val tree2 = tree.withType(defn.RootPackage.termRef)
+      checkLegalValue(tree2, pt)
+      return tree2
 
     val rawType =
       val saved1 = unimported
@@ -581,13 +583,13 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
         if foundUnderScala2.exists && !(foundUnderScala2 =:= found) then
           report.migrationWarning(
             em"""Name resolution will change.
-              | currently selected                          : $foundUnderScala2
-              | in the future, without -source 3.0-migration: $found""", tree.srcPos)
+                | currently selected                          : $foundUnderScala2
+                | in the future, without -source 3.0-migration: $found""", tree.srcPos)
           foundUnderScala2
         else found
       finally
-      	unimported = saved1
-      	foundUnderScala2 = saved2
+        unimported = saved1
+        foundUnderScala2 = saved2
 
     /** Normally, returns `ownType` except if `ownType` is a constructor proxy,
      *  and there is another shadowed type accessible with the same name that is not:

--- a/tests/neg/i18020.scala
+++ b/tests/neg/i18020.scala
@@ -1,0 +1,57 @@
+import _root_.scala.StringContext // ok
+
+class Test :
+  val Foo = 1
+  def foo0: Unit =
+    val x = new _root_.scala.StringContext() // ok
+    val y: Option[_root_.scala.Serializable] = None // ok
+    val z: _root_.scala.None.type = None
+    val w = _root_.scala.None
+    val (_root_, other) = (1, 2) // error
+    val (Test.this.Foo, 1) = ???
+    ??? match
+      case (Test.this.Foo, 1) => ()
+
+def foo3 =
+  val _root_ = "abc" // error
+
+def foo1: Unit =
+  val _root_: String = "abc" // error // error
+  // _root_: is, technically, a legal name
+  // so then it tries to construct the infix op pattern
+  // "_root_ String .." and then throws in a null when it fails
+  // to find an argument
+  // then Typer rejects "String" as an infix extractor (like ::)
+  // which is the second error
+
+def foo2: Unit = // error
+  val _root_ : String = "abc" // error
+
+// i17757
+def fooVal: Unit =
+  val _root_  = "abc" // error
+  println(_root_.length)  // error
+  println(_root_)  // error
+
+def barVal: Unit =
+  _root_  // error
+  _root_.scala  // error
+  println(_root_)  // error
+  println(_root_.scala)  // error
+
+// i18050
+package p {
+  package _root_ { // error
+    object X // error
+  }
+}
+
+// i12508
+package _root_ { // error
+  class C {
+    val _root_ = 42 // error
+  }
+}
+package _root_.p { // error
+  class C
+}

--- a/tests/neg/i18020.scala
+++ b/tests/neg/i18020.scala
@@ -46,7 +46,7 @@ package p {
   }
 }
 
-// i12508
+// scala/bug#12508
 package _root_ { // error
   class C {
     val _root_ = 42 // error
@@ -55,3 +55,8 @@ package _root_ { // error
 package _root_.p { // error
   class C
 }
+
+// from ScalaPB
+def fromScalaPb(x: Option[String]) = x match
+  case _root_.scala.Some(s) => s
+  case _                    => ""

